feat(ras-acc): dynamically set otp email text color

chickenn00dle opened the pull request:
### All Submissions:

* [ ] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/trunk/.github/CONTRIBUTING.md)?
* [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request, and the reason for such changes. -->

**PR 2 of 4**

This PR adds onto the work in https://github.com/Automattic/newspack-plugin/pull/3032 by adjusting email text, button, and social icon color for the default email template to either white or black depending on background color:

![Screenshot 2024-04-04 at 17 35 31](https://github.com/Automattic/newspack-plugin/assets/17905991/d5a25d21-e325-4d12-8206-4a91c54a940c)

Note: The changes only apply to the one time password email for the time being. Once that email is fully updated, I will get a PR up to update the remaining templates.

### How to test the changes in this Pull Request:

1. Go to the WP Customizer and set a dark custom primary and secondary color
2. With these changes checked out, go to Newspack > Engagment > Show Advanced Settings and select the one-time password email under Transactional Email Content section
3. Trash the post (so that we can force our updated template to be used on next otp sign in attempt)
4. As a logged out user, select the sign in link and attempt to login via email to trigger an otp email
5. Check the otp email and confirm all of the text is properly contrasting
6. Repeat the above steps with light custom colors in customizer

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

chickenn00dle committed:
feat(ras-acc): update otp email template

chickenn00dle committed:
feat(ras-acc): add social icons footer to transactional emails

chickenn00dle committed:
fix(ras-acc): fix otp email formatting and translations

chickenn00dle committed:
fix(ras-acc): email template bg color newsletter conflict

chickenn00dle committed:
feat(ras-acc): make otp email colors dynamic

chickenn00dle commented on includes/util.php:
```diff
@@ -495,16 +495,78 @@ function newspack_get_countries() {
+ *
+ * @ref https://stackoverflow.com/questions/1331591/given-a-background-color-black-or-white-text
+ */
+function newspack_get_color_contrast( $hex ) {
```
This utility function lives in newspack theme, but given not every publisher will be using this theme, and given how useful this function is, I've copied it over to the main newspack plugin. 

I'm happy to move it back if someone feels it shouldn't be copied over though!

chickenn00dle committed:
fix: remove redundant code from rebase

adekbadek commented on includes/templates/reader-activation-emails/otp.php:
```diff
@@ -9,49 +9,54 @@
-	<div class="wp-block-buttons has-custom-font-size has-medium-font-size" style="font-style:normal;font-weight:400"><!-- wp:button {"textAlign":"center","style":{"color":{"background":"' . esc_attr( $primary_color ) . '"},"border":{"radius":"4px"}}} -->
-	<div class="wp-block-button"><a class="wp-block-button__link has-background has-text-align-center wp-element-button" href="*MAGIC_LINK_URL*" style="border-radius:4px;background-color:' . esc_attr( $primary_color ) . '">' . sprintf( /* Translators: s: site title. */ __( 'Continue to %s', 'newspack' ), '*SITE_TITLE*' ) . '</a></div>
+	<div class="wp-block-buttons has-custom-font-size has-medium-font-size" style="font-style:normal;font-weight:400"><!-- wp:button {"textAlign":"center","textColor":"' . esc_attr( $primary_text_color ) . '","style":{"color":{"background":"' . esc_attr( $primary_color ) . '"},"border":{"radius":"4px"},"elements":{"link":{"color":{"text":"' . esc_attr( $primary_text_color ) . '"}}}}} -->
+	<div class="wp-block-button"><a class="wp-block-button__link has-' . esc_attr( $primary_text_color ) . '-color has-text-color has-background has-link-color has-text-align-center wp-element-button" href="*MAGIC_LINK_URL*" style="border-radius:4px;background-color:' . esc_attr( $primary_color ) . '">Continue to *SITE_TITLE*</a></div>
```
Is the removal of translation function (`__()`) intentional here?

chickenn00dle replied:
This was not intentional. Fixed in https://github.com/Automattic/newspack-plugin/pull/3033/commits/6510cc350130095d99547fc49098875bdd880ede

adekbadek commented on includes/templates/reader-activation-emails/otp.php:
```diff
@@ -9,49 +9,54 @@
 	<!-- /wp:social-links -->
-
 	<!-- wp:paragraph {"fontSize":"small"} -->
 	<p class="has-small-font-size">' . sprintf( /* Translators: 1: site title 2: site base address. */ __( '%1$s - %2$s', 'newspack' ), '<strong>*SITE_TITLE*</strong>', '*SITE_ADDRESS*' ) . '<br>' . sprintf( /* Translators: 1: link to site url. */ __( 'You received this email because you requested to sign in to %s', 'newspack' ), '<a href="*SITE_URL*">*SITE_URL*</a>' ) . '</p>
```
I've just noticed – the text domain in the translation functions is wrong. Should be `'newspack-plugin'`.

chickenn00dle replied:
Good catch! Fixed in https://github.com/Automattic/newspack-plugin/pull/3033/commits/6510cc350130095d99547fc49098875bdd880ede

adekbadek reviewed (changes requested):
Two issues with the template file:
1. the `$email_html` is not updated along with the `$post_content` changes
2. there's a hard-coded OTP code with `895671` value (instead of a placeholder)

chickenn00dle committed:
fix: typos and translations

chickenn00dle committed:
feat(ras-acc): update colors on theme color update

chickenn00dle committed:
fix: display correct text colors in rr emails

chickenn00dle commented:
Thanks for the review @adekbadek!

> the $email_html is not updated along with the $post_content changes

I've updated the email html and did some formatting to hopefully make parsing the email a bit easier.

> there's a hard-coded OPT code with 895671 value (instead of a placeholder)

Good catch. Fixed in https://github.com/Automattic/newspack-plugin/pull/3033/commits/6510cc350130095d99547fc49098875bdd880ede

I also made some additional changes to how we set colors in the default block in https://github.com/Automattic/newspack-plugin/pull/3033/commits/e574d6c8d2ba63ebc42fcd62f87e40188d855c6e and https://github.com/Automattic/newspack-plugin/pull/3033/commits/e4c84731219bd1c3a170fcbb016e93acd061f427. I realized the previous method would not allow the email template to update when theme colors were changed since the block colors were being explicitly set as hex values. 

Unfortunately, it looks like we will not be able to use a dynamic approach for social icons without some changes in the newsletter plugin first since newsletters pretty strictly decides on a colorsheme for social icon images when generating the email html markup. (at least I wasn't able to find a straightforward way without having to make changes in newsletters)

So I've reverted the icons to use the default layout and colorscheme for the time being. I will try to address them in a follow-up PR once I've worked out all of the other design bits for emails:

![Screenshot 2024-04-16 at 18 30 13](https://github.com/Automattic/newspack-plugin/assets/17905991/2ebbed05-f696-4023-a01a-249c719067be)

adekbadek commented on includes/util.php:
```diff
@@ -495,49 +495,116 @@ function newspack_get_countries() {
+
 	if ( ! is_wp_error( $cm ) ) {
 		$social_urls = [
+			// TODO: Add more social media services here.
```
Why not add them now? From what I see [here](https://github.com/Automattic/newspack-plugin/blob/8b5d5c4c8b7372dd33603988142097435997bf2a/includes/class-profile.php#L36-L69), only LinkedIn and Pinterest are missing.

chickenn00dle replied:
Fair point! I've added linkedin in https://github.com/Automattic/newspack-plugin/pull/3033/commits/ff88d4b851c20019c59edacd517f9237a95f0f53

It looks like newsletters doesn't support pinterest though so I've omitted that one for the time being:
https://github.com/Automattic/newspack-newsletters/blob/6711238e09a20dd215833207810e9e341cd54a3f/includes/class-newspack-newsletters-renderer.php#L164-L173

I've added this to the list of small changes needed in newsletters and will tackle this altogether in a follow-up PR: 0-as-1205668937699542/1207101310249278/f

adekbadek commented on includes/templates/reader-activation-emails/otp.php:
```diff
@@ -9,217 +9,278 @@
 	<!-- wp:buttons {"layout":{"type":"flex","orientation":"horizontal","flexWrap":"nowrap","justifyContent":"left"},"style":{"typography":{"fontStyle":"normal","fontWeight":"400"}},"fontSize":"medium"} -->
-	<div class="wp-block-buttons has-custom-font-size has-medium-font-size" style="font-style:normal;font-weight:400"><!-- wp:button {"textAlign":"center","style":{"color":{"background":"' . esc_attr( $primary_color ) . '"},"border":{"radius":"4px"}}} -->
-	<div class="wp-block-button"><a class="wp-block-button__link has-background has-text-align-center wp-element-button" href="*MAGIC_LINK_URL*" style="border-radius:4px;background-color:' . esc_attr( $primary_color ) . '">' . sprintf( /* Translators: s: site title. */ __( 'Continue to %s', 'newspack' ), '*SITE_TITLE*' ) . '</a></div>
+	<div class="wp-block-buttons has-custom-font-size has-medium-font-size" style="font-style:normal;font-weight:400"><!-- wp:button {"textAlign":"center","backgroundColor":"primary","textColor":"secondary","style":{"border":{"radius":"4px"},"elements":{"link":{"color":{"text":"secondary"}}}}} -->
```
Is setting the text color to the secondary color intentional? On the screenshot in PR description it's the contrasting text color, not theme secondary color:

<img width="589" alt="image" src="https://github.com/Automattic/newspack-plugin/assets/7383192/76ac9e3b-dead-435b-a1eb-c64029c4b08c">

chickenn00dle replied:
> Is setting the text color to the secondary color intentional? On the screenshot in PR description it's the contrasting text color, not theme secondary color

Sorry about that. It is. I had another look at the designs (c6T1goPQpkchdh7XBAM6aH-fi-1_10676) and it seems like the button text is meant to be the secondary color.

chickenn00dle committed:
feat(ras-acc): add other social services to email footer
